### PR TITLE
Enable production queue additions from object wnd

### PR DESCRIPTION
**Pull request by dbenage-cx**
Adds right click contexts for planets in the object window, adding ship designs or buildings to the production queue.

### CODE REVIEW

File: UI/ObjectListWnd.cpp [hunk 1]
```diff
@@ -2424,6 +2430,20 @@ void ObjectListWnd::ObjectRightClicked(GG::ListBox::iterator it, const GG::Pt& p
                     std::vector<std::string> planet_foci = one_planet->AvailableFoci();
                     for (std::vector<std::string>::iterator it = planet_foci.begin(); it != planet_foci.end(); ++it)
                         all_foci[*it]++;
+
+                    std::set<int> shipDesigns = cur_empire->AvailableShipDesigns();
```
Comment (geoffthemedio) [inline UI/ObjectListWnd.cpp]: shipDesigns -> ship_designs

File: UI/ObjectListWnd.cpp [hunk 2]
```diff
@@ -2424,6 +2430,20 @@ void ObjectListWnd::ObjectRightClicked(GG::ListBox::iterator it, const GG::Pt& p
...
+                    for (std::set<int>::iterator sd_it = shipDesigns.begin(); sd_it != shipDesigns.end(); ++sd_it) {
+                        if (cur_empire->ProducibleItem(BT_SHIP, *sd_it, row->ObjectID()))
+                            avail_designs[*sd_it]++;
+                    }
+
+                    const std::set<std::string>& buildingTypes = cur_empire->AvailableBuildingTypes();
```
Comment (geoffthemedio) [inline UI/ObjectListWnd.cpp]: buildingTypes -> building_types

File: UI/ObjectListWnd.cpp [hunk 3]
```diff
@@ -2436,6 +2456,28 @@ void ObjectListWnd::ObjectRightClicked(GG::ListBox::iterator it, const GG::Pt& p
         }
         if (menuitem_id > MENUITEM_SET_FOCUS_BASE)
             menu_contents.next_level.push_back(focusMenuItem);
+
+        GG::MenuItem shipDesignMenuItem(UserString("MENUITEM_ENQUEUE_SHIPDESIGN"), 4, false, false);
```
Comment (geoffthemedio) [inline UI/ObjectListWnd.cpp]: etc...

File: UI/ObjectListWnd.cpp [hunk 4]
```diff
@@ -2409,9 +2409,15 @@ void ObjectListWnd::ObjectRightClicked(GG::ListBox::iterator it, const GG::Pt& p
         return;
 
     const int MENUITEM_SET_FOCUS_BASE = 20;
+    const int MENUITEM_SET_SHIP_BASE = 50;
+    const int MENUITEM_SET_BUILDING_BASE = 100;
```
Comment (geoffthemedio) [inline UI/ObjectListWnd.cpp]: need to space out these numbers a lot more for user-customizable stuff... a player could easily have more than 50 producible ship designs. alternatively, add some limits so never more than the 50 most commonly-used ship designs are listed in the droplist.

File: default/stringtables/en.txt [hunk 5]
```diff
@@ -2655,6 +2655,12 @@ Latest known [[encyclopedia STEALTH_TITLE]] is out of date. Actual stealth excee
 MENUITEM_SET_FOCUS
 Set focus
 
+MENUITEM_ENQUEUE_BUILDING
+Queue Building
+
+MENUITEM_ENQUEUE_SHIPDESIGN
+Queue Ship Design
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: just "Queue Ship"

File: UI/ObjectListWnd.cpp [hunk 6]
```diff
@@ -2485,7 +2527,42 @@ void ObjectListWnd::ObjectRightClicked(GG::ListBox::iterator it, const GG::Pt& p
...
+                    }
+                }
+            } else if (id > MENUITEM_SET_BUILDING_BASE && id <= bld_menuitem_id) {
+                std::map<std::string, int>::iterator it = avail_blds.begin();
+                std::advance(it, id - MENUITEM_SET_BUILDING_BASE - 1);
+                std::string bld = it->first;
+                const GG::ListBox::SelectionSet sel = m_list_box->Selections();
+                for (GG::ListBox::SelectionSet::const_iterator it = sel.begin(); it != sel.end(); ++it) {
+                    ObjectRow *row = dynamic_cast<ObjectRow *>(**it);
+                    if (row) {
```
Comment (geoffthemedio) [inline UI/ObjectListWnd.cpp]:
```
if (!row) continue
```

and avoid unnecessary indentation below

File: UI/ObjectListWnd.cpp [hunk 7]
```diff
@@ -2485,7 +2527,42 @@ void ObjectListWnd::ObjectRightClicked(GG::ListBox::iterator it, const GG::Pt& p
...
+                        TemporaryPtr<Planet> one_planet = GetPlanet(row->ObjectID());
+                        if (one_planet && one_planet->OwnedBy(app->EmpireID())) {
```
Comment (geoffthemedio) [inline UI/ObjectListWnd.cpp]:
```
if (!one_planet || !one_planet->OwnedBy(...)) continue
```

(etc.)

**Comment by MatGB** [comment]
Compiles fine on my machine, observations: appears to list all buildings unlocked for every planet, regardless of whether they can be built there, eg a black hole power generator showed up on my neutron star gas giant (that I queued an artificial paradise world on to test).

Also, I queued a Chato colony using this, but it showed as 'never complete' in my production window, deleted it and added it traditional way and it was then showing 8 turns. Haven't replicated that yet but only just started playing with it.

**Comment by dbenage-cx** [comment]
Rebased addressing Geoff's notes and updating production queue(never complete).

@MatGB Indeed, some buildings are passing the Producable/Enqueable check that should not, e.g. Drydocks.  Will see if I can track it down tonight.

**Comment by dbenage-cx** [comment]
Missed ProducibleItem check for buildings.

File: UI/ObjectListWnd.cpp [hunk 8]
```diff
@@ -2436,6 +2457,29 @@ void ObjectListWnd::ObjectRightClicked(GG::ListBox::iterator it, const GG::Pt& p
         }
         if (menuitem_id > MENUITEM_SET_FOCUS_BASE)
             menu_contents.next_level.push_back(focusMenuItem);
+
+        GG::MenuItem ship_menu_item(UserString("MENUITEM_ENQUEUE_SHIPDESIGN"), 4, false, false);
+        for (std::map<int, int>::iterator it = avail_designs.begin();
+             it != avail_designs.end() && ship_menuitem_id < MENUITEM_SET_BUILDING_BASE; ++it) {
```
Comment (geoffthemedio) [inline UI/ObjectListWnd.cpp]: { to next line

File: UI/ObjectListWnd.cpp [hunk 9]
```diff
@@ -2424,6 +2430,21 @@ void ObjectListWnd::ObjectRightClicked(GG::ListBox::iterator it, const GG::Pt& p
...
+                            avail_designs[*sd_it]++;
+                    }
+
+                    const std::set<std::string>& building_types = cur_empire->AvailableBuildingTypes();
+                    for (std::set<std::string>::const_iterator bld_it = building_types.begin();
+                         bld_it != building_types.end(); ++bld_it)
+                    {
+                        if (cur_empire->EnqueuableItem(BT_BUILDING, *bld_it, row->ObjectID()) &&
+                            cur_empire->ProducibleItem(BT_BUILDING, *bld_it, row->ObjectID()))
+                            avail_blds[*bld_it]++;
```
Comment (geoffthemedio) [inline UI/ObjectListWnd.cpp:2447]: After a multi-line if, put some {} around the statement

**Comment by geoffthemedio** [comment]
Seems to work.

**Comment by dbenage-cx** [comment]
Rebased

**Pull request merged**